data_pipeline/generic_web_api/generic_web_api_data_etl.py

```diff
@@ -0,0 +1,448 @@
+        from_date=from_date,
+        to_date=until_date,
+        cursor=cursor,
+        page_number=page_number,
+    )
+
+    with requests_retry_session() as session:
+        if (
+                data_config.authentication and
+                data_config.authentication.authentication_type == "basic"
+        ):
+            session.auth = tuple(data_config.authentication.auth_val_list)
+        session.verify = False
+        session_request = session.get(url)
+        session_request.raise_for_status()
+        resp = session_request.content
+        try:
+            json_resp = json.loads(resp)
+        except JSONDecodeError:
+            json_resp = get_newline_delimited_json_string_as_json_list(
+                resp.decode("utf-8")
+            )
+    return json_resp
+
+
+def generic_web_api_data_etl(
+        data_config: WebApiConfig,
+        full_temp_file_location,
+        from_date: datetime = None,
+        until_date: datetime = None,
+):
+    stored_state = get_stored_state(
+        data_config
+    )
+
+    from_date = from_date if from_date else stored_state
+    cursor = None
+    imported_timestamp = get_current_timestamp_as_string()
+    latest_record_timestamp = None
+    page_number = 1 if data_config.url_manager.page_number_param else None
+    while True:
+        page_data = get_data_single_page(
+            data_config=data_config,
+            from_date=from_date,
+            until_date=until_date,
+            cursor=cursor,
+            page_number=page_number,
+        )
+        items_list = get_items_list(
+            page_data, data_config
+        )
+        latest_record_timestamp = process_downloaded_data(
+            data_config=data_config,
+            record_list=items_list,
+            data_etl_timestamp=imported_timestamp,
+            file_location=full_temp_file_location,
+            prev_page_latest_timestamp=latest_record_timestamp
+        )
+        items_count = len(items_list)
+        cursor = get_next_cursor_from_data(page_data, data_config)
+        page_number = get_next_page_number(
+            items_count, page_number, data_config
+        )
+        from_date = get_next_start_date(
+            items_count, from_date,
+            latest_record_timestamp, data_config
+        )
+        if cursor is None and page_number is None and from_date is None:
+            break
+
+    create_n_extend_table(data_config, full_temp_file_location)
+
+    load_file_into_bq(
+        filename=full_temp_file_location,
+        table_name=data_config.table_name,
+        auto_detect_schema=False,
+        dataset_name=data_config.dataset_name,
+        project_name=data_config.gcp_project,
+    )
+    upload_latest_timestamp(
+        data_config, latest_record_timestamp
+    )
+
+
+def get_next_page_number(items_count, current_page, web_config: WebApiConfig):
+    next_page = None
+    if web_config.url_manager.page_number_param:
+        to_continue = (
```

Comment:
minor: just wondering whether something like `has_more_items` would be more intuitive?